src/lib/communication/DY.Lib.Communication.API.Predicates.fst

```diff
@@ -0,0 +1,138 @@
+        get_sk_label pk == principal_label receiver /\
+        (get_label payload) `can_flow tr` (join (principal_label sender) (principal_label receiver)) /\
+        event_triggered tr sender (CommConfSendMsg sender receiver payload)
+      )
+      | _ -> False)
+    );
+  pred_later = (fun tr1 tr2 pk msg -> ());
+}
+#pop-options
+
+val pkenc_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & pkenc_crypto_predicate cusages)
+let pkenc_crypto_predicates_communication_layer_and_tag #cusages = 
+  (comm_layer_pkenc_tag, pkenc_crypto_predicates_communication_layer cusages)
+
+(*** Sign Predicates ***)
+
+#push-options "--ifuel 3 --fuel 0"
+val sign_crypto_predicates_communication_layer: cusages:crypto_usages -> sign_crypto_predicate cusages
+let sign_crypto_predicates_communication_layer cusages = {
+  pred = (fun tr vk sig_msg ->
+    get_signkey_usage vk == SigKey comm_layer_sign_tag empty /\
+    (match parse communication_message sig_msg with
+    | Some (CM {sender; receiver; payload}) -> (
+      get_signkey_label vk == principal_label sender /\
+      get_label payload `can_flow tr` public /\
+      event_triggered tr sender (CommAuthSendMsg sender payload)
+    )
+    | _ -> False)
+  );
+  pred_later = (fun tr1 tr2 vk msg -> ());
+}
+#pop-options
+
+val sign_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & sign_crypto_predicate cusages)
+let sign_crypto_predicates_communication_layer_and_tag #cusages =
+  (comm_layer_sign_tag, sign_crypto_predicates_communication_layer cusages)
+
+val has_communication_layer_invariants:
+  crypto_invariants ->
+  prop
+let has_communication_layer_invariants cinvs =
+  has_pkenc_predicate cinvs pkenc_crypto_predicates_communication_layer_and_tag /\
+  has_sign_predicate cinvs sign_crypto_predicates_communication_layer_and_tag
+
+(*** Event Predicates ***)
+
+noeq
+type comm_higher_layer_event_preds = {
+  send_conf: sender:principal -> receiver:principal -> payload:bytes -> tr:trace -> prop;
+  // TODO do we need extensibility on the receiver side?
+  receive_conf: sender:principal -> receiver:principal -> payload:bytes -> tr:trace -> prop;
+  send_auth: sender:principal -> payload:bytes -> tr:trace -> prop;
+  // TODO do we need extensibility on the receiver side?
+  receive_auth: sender:principal -> receiver:principal -> payload:bytes -> tr:trace -> prop;
+}
+
+val default_comm_higher_layer_event_preds: comm_higher_layer_event_preds
+let default_comm_higher_layer_event_preds = {
+  send_conf = (fun tr sender receiver payload -> True);
+  receive_conf = (fun tr sender receiver payload -> True);
+  send_auth = (fun tr sender payload -> True);
+  receive_auth = (fun tr sender receiver payload -> True);
+}
+
+val comm_higher_layer_event_preds_later: trace -> (trace -> prop) -> prop
+let comm_higher_layer_event_preds_later tr higher_layer_preds =
+  forall tr'. tr <$ tr' ==> higher_layer_preds tr'
```

Comment:
Does it happen that higher layer predicates are not preserved by trace extension? It seems to me that in all preconditions of the `Lemmas` file use `comm_higher_layer_event_preds_later`, so it might make sense to move this fact in `comm_higher_layer_event_preds`. When doing that we only need to prove the monotonicity of predicates when defining them (once and for all), whereas here we would need to prove it each time we use a lemma that uses `comm_higher_layer_event_preds_later` in the precondition.